src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -1268,39 +1268,47 @@ public DynamicObject replace(DynamicObject string, DynamicObject other) {
     @ImportStatic(StringGuards.class)
     public abstract static class RstripBangNode extends CoreMethodArrayArgumentsNode {
 
+        @Child private RopeNodes.GetCodePointNode getCodePointNode = RopeNodes.GetCodePointNode.create();
         @Child private RopeNodes.MakeSubstringNode makeSubstringNode = RopeNodes.MakeSubstringNode.create();
 
         @Specialization(guards = "isEmpty(string)")
         public DynamicObject rstripBangEmptyString(DynamicObject string) {
             return nil();
         }
 
-        @TruffleBoundary
         @Specialization(guards = { "!isEmpty(string)", "isSingleByteOptimizable(string)" })
-        public Object rstripBangSingleByte(DynamicObject string) {
+        public Object rstripBangSingleByte(DynamicObject string,
+                                           @Cached("create()") RopeNodes.BytesNode bytesNode,
+                                           @Cached("createBinaryProfile()") ConditionProfile noopProfile) {
             // Taken from org.jruby.RubyString#rstrip_bang19 and org.jruby.RubyString#singleByteRStrip19.
 
             final Rope rope = rope(string);
-            final byte[] bytes = rope.getBytes();
-            final int start = 0;
-            final int end = rope.byteLength();
-            int endp = end - 1;
-            while (endp >= start && (bytes[endp] == 0 ||
-                    ASCIIEncoding.INSTANCE.isSpace(bytes[endp] & 0xff))) endp--;
+            final int lastCodePoint = getCodePointNode.executeGetCodePoint(rope, rope.byteLength() - 1);
+
+            // Check the last code point to see if it's a space or NULL. In the case of strings without leading spaces,
+            // this check can avoid having to materialize the entire byte[] (a potentially expensive operation
+            // for ropes) and can avoid having to compile the while loop.
+            final boolean willStrip = lastCodePoint == 0x00 || StringSupport.isAsciiSpace((byte) lastCodePoint);
```

Comment:
It's odd rstrip! removes \0 but lstrip! does not, but that seems MRI semantics.
The RDoc documention of these methods doesn't say anything about it.

Reply:
Yeah. I'm really not sure why that is.